sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala

```diff
@@ -351,10 +351,20 @@ object PartitioningUtils {
    */
   def getPathFragment(spec: TablePartitionSpec, partitionSchema: StructType): String = {
     partitionSchema.map { field =>
-      escapePathName(field.name) + "=" + getPartitionValueString(spec(field.name))
+      escapePathName(field.name) + "=" +
+        getPartitionValueString(
+          removeLeadingZerosFromNumberTypePartition(spec(field.name), field.dataType))
     }.mkString("/")
   }
 
+  def removeLeadingZerosFromNumberTypePartition(value: String, dataType: DataType): String = {
```

Comment:
I didn't take a close look but I do think we should deduplicate these codes. At the very least you can reuse cast expression

Reply:
This comment is still outstanding - any possibility to refactor?

Reply:
I tried to refactor and add an unit test. Could you please review it again? Thank you.